Домашнее задание 7. Департамент ATM

LovchikovaES opened the pull request.

LovchikovaES committed:
Домашнее задание 7. Департамент ATM

spvgl commented on hw07-ATMDepartment/src/test/java/ru/catn/ATMDepartmentTest.java:
```diff
@@ -0,0 +1,90 @@
+                case 2:
+                    notesToPut.put(noteOf1000, 2);
+                    break;
+            }
```
Здесь удобно сделать какой-нибудь builder или, скорее, factory для создания `Note` по целочисленному значению.

LovchikovaES replied:
Вы имеете ввиду типа такой factory: в атрибутах мапа, а в методах putNoteOf1000(int quantity)?

spvgl replied:
Нет. Что-то вроде этого:
```
for (var atm : atms) {
    Note note = notesFactory.getNote(i);
    notesToPut.put(note, 2);
    ...
}
```
Хотя, если  такое будет только в тестах, то скорее лишнее.

LovchikovaES replied:
М, этот цикл у меня был, чтобы заполнить разные банкоматы купюрами. Не совсем понимаю как notesFactory избавит меня от switch-a? 

Или это не про switch, а про вот такую запись: notesToPut.put(noteOf1000, 2) ? То есть мы получим банкноту по номиналу тут (Note note = notesFactory.getNote(i)), верно? И где такую factory лучше хранить? В atm? И в конструкторе её заполнить, верно?

spvgl replied:
Да, это про noteOf1000.
Фабрику отдельным классом сделать

spvgl commented on hw07-ATMDepartment/src/main/java/ru/catn/cell/CellHistory.java:
```diff
@@ -0,0 +1,19 @@
+    private Deque<CellState> states = new ArrayDeque<>();
+
+    public void addState(CellState state) {
+        states.add(state);
```
Для надежности я бы создавал здесь новый `CellState` (делал копию `state`). 
Да, в другом месте (перед вызовом `addState()` в `cell.getState())`) у вас создается новый `CellState` и работает все правильно, но можно забыть передать сюда именно новый `CellState`. Лучше инкапсулировать эту логику здесь.

LovchikovaES replied:
Спасибо, поправила

spvgl replied:
Принято

spvgl commented on hw07-ATMDepartment/src/main/java/ru/catn/RestoreATMProducer.java:
```diff
@@ -0,0 +1,16 @@
+        listeners.add(listener);
+    }
+
+    void event() {
```
Метод лучше назвать глаголом, что-то вроде, `processEvents` или `restore`.

LovchikovaES replied:
Спасибо, поправила

spvgl replied:
Принято

LovchikovaES committed:
Домашнее задание 7. Департамент ATM. Исправления:
1) изменено название метода в RestoreATMProducer
2) при добавлении состояния в CellHistory создаётся копия состояния

LovchikovaES committed:
Домашнее задание 7. Департамент ATM. Создана фабрика для банкнот и реорганизован каталог классов по пакетам
